### PR TITLE
Workaround v8 bug

### PR DESCRIPTION
**Pull request by nsavoire**
v8 profiler event processor thread sometimes get stuck in a loop eating 100% CPU, leading to empty profiles.
This pull-request tries to accomplish goals:
* Make the bug less likely to happen by waiting for profiler signal before and after starting a new profile
* Detect if the bug has occurred (by checking for profiles with no hits or no non-ticks samples), and in that case fully stop and restart the profiler to reset the profile to a valid state.

### CODE REVIEW

File: ts/src/time-profiler.ts [hunk 1]
```diff
@@ -141,5 +168,9 @@ export function isStarted() {
   return !!gProfiler;
 }
 
+export function v8ProfilerStuckEventoopDetected() {
```
Comment (szegedi) [inline ts/src/time-profiler.ts:171]:
```suggestion
export function v8ProfilerStuckEventLoopDetected() {
```

File: ts/src/index.ts [hunk 2]
```diff
@@ -35,6 +35,7 @@ export const time = {
   stop: timeProfiler.stop,
   setContext: timeProfiler.setContext,
   isStarted: timeProfiler.isStarted,
+  v8ProfilerStuckEventoopDetected: timeProfiler.v8ProfilerStuckEventoopDetected,
```
Comment (szegedi) [inline ts/src/index.ts:38]:
```suggestion
  v8ProfilerStuckEventLoopDetected: timeProfiler.v8ProfilerStuckEventLoopDetected,
```

**Review by szegedi — CHANGES_REQUESTED** [review]
Just an unfortunate typo. Otherwise looks good!

File: ts/src/time-profiler.ts (hunk 1 above)
Comment (nsavoire) [reply ts/src/time-profiler.ts]: Woops, thanks for the review !
I added some tests :)

File: bindings/profilers/wall.cc [hunk 3]
```diff
@@ -54,6 +54,29 @@ using namespace v8;
...
+}
+
+bool detectV8Bug(const v8::CpuProfile* profile) {
+  // Return true if either condition is false:
+  // - profile has at least one node with a non-zero hitcount
+  // - number of samples is strictly greater to total hitcount (meaning that
+  //   starting sample or some deopt events have been processed)
+  // If either condition is false, it implies that
+  // v8::SamplingEventsProcessor::ProcessOneSample loop is stuck for
+  // ticks_buffer_ or vm_ticks_buffer_.
```
Comment (szegedi) [inline bindings/profilers/wall.cc:74]: It's probably me, but I find it a bit convoluted to think in terms of the NAND operator ("true if either is false") :-) Can you rephrase it?
Here's a suggestion:
> When the profiler operates correctly, there'll be at least one node with positive hitcount and the number of samples will be strictly greater than the number of hits because they'll contain at least the starting sample and potentially some deoptimization samples. If these conditions don't hold, it implies that…

File: bindings/profilers/wall.cc [hunk 4]
```diff
@@ -54,6 +54,29 @@ using namespace v8;
...
+
+  auto totalHitCount = getTotalHitCount(profile->GetTopDownRoot());
+  return totalHitCount == 0 || profile->GetSamplesCount() == totalHitCount;
```
Comment (szegedi) [inline bindings/profilers/wall.cc:77]: Maybe just to be on the safe side?
```suggestion
  return totalHitCount == 0 || profile->GetSamplesCount() <= totalHitCount;
```

**Review by szegedi — CHANGES_REQUESTED** [review]

File: bindings/profilers/wall.cc [hunk 5]
```diff
@@ -489,8 +489,11 @@ Result WallProfiler::StartImpl() {
 }
 
 std::string WallProfiler::StartInternal() {
+  // Reuse the same names for the profiles because strings used for profile
+  // names are not released until v8::CpuProfiler object is destroyed.
+  // https://github.com/nodejs/node/blob/b53c51995380b1f8d642297d848cab6010d2909c/deps/v8/src/profiler/profile-generator.h#L516
   char buf[128];
-  snprintf(buf, sizeof(buf), "pprof-%" PRId64, profileIdx_++);
+  snprintf(buf, sizeof(buf), "pprof-%" PRId64, (profileIdx_++) % 2);
```
Comment (szegedi) [inline bindings/profilers/wall.cc:496]: I have to admit I don't understand this.

File: bindings/profilers/wall.cc (hunk 4 above)
Comment (nsavoire) [reply bindings/profilers/wall.cc:77]: I think it's better to check for equality, if sample count is strictly lower, it implies that tick samples were dropped and this is very likely a false positive.

File: bindings/profilers/wall.cc (hunk 5 above)
Comment (nsavoire) [reply bindings/profilers/wall.cc:496]: As discussed, modulo operation allows to alternate between "pprof-0" and "pprof-1",  we know that there are most 2 profiles active at the same time.

File: bindings/profilers/wall.cc [hunk 6]
```diff
@@ -553,14 +602,44 @@ NAN_METHOD(WallProfiler::Stop) {
...
+    // no target call count in input, wait for the next signal
+    targetCallCount = currentCallCount + 1;
+  }
+#if DD_WALL_USE_SIGPROF
+  const int maxRetries = 2;
+  // wait for a maximum of 2 sample period
+  // if a signal occurs it will interrupt sleep (we use nanosleep and not
+  // uv_sleep because we want this behaviour)
+  timespec ts = {0, samplingPeriodMicros_ * maxRetries * 1000};
+  nanosleep(&ts, nullptr);
```
[the remark this answers is not in the document]
Comment (nsavoire) [reply bindings/profilers/wall.cc:623]: Yes indeed, that's the annoying side effect of the workaround. But as you mentioned, 20ms every minute is ok I guess.

**Review by szegedi — APPROVED** [review]
